std/stdio.d

```diff
@@ -4043,35 +4082,40 @@ private struct ReadlnAppender
     import core.stdc.string;
 
     char[] buf;
-    size_t cap;
     size_t pos;
-    size_t initcap;
+    bool newBuf = false;
 
     void initialize(char[] b)
     {
         buf = b;
-        cap = initcap = buf.capacity;
-        if (cap < buf.length)
-            cap = buf.length;
         pos = 0;
     }
     @property char[] data()
     {
-        // always shrink/extend the buffer to allow reuse in the next call
-        if (initcap > 0 || cap > buf.length)
+        /* When buf is not known to the outside world, it's safe to append to
+        the returned slice. */
+        if (newBuf)
             assumeSafeAppend(buf.ptr[0..pos]);
         return buf.ptr[0..pos];
     }
 
     void reserve(size_t n)
     {
-        if (pos + n > cap)
+        if (buf.length >= pos + n) // buf is already large enough
+            return;
+
+        immutable curCap = buf.capacity;
+        if (curCap >= pos + n)
+        {
+            buf.length = curCap;
```

Comment:
Sure.

I look at it this way:

1. `buffer` is not big enough to hold the data, so we must expand
2. Since `buffer.capacity` is non-zero, we know (or we can safely assume) that the data after `buffer` is unused.
3. When we expand `buffer` to its capacity, it's likely we will not use all the data, so the extra data we claim here for expanded capacity will be garbage if we don't `assumeSafeAppend`.

So if you set `newBuf = true`, then `assumeSafeAppend` will be called at the end, and the data we claimed but didn't use is still available for appending. Effectively it's the same thing as your original implementation, but without the capacity check in the inner loop.